src/tools/ceph_objectstore_tool.cc

```diff
@@ -1444,20 +1461,52 @@ int do_remove_object(ObjectStore *store, coll_t coll,
     return r;
   }
 
-  cout << "remove " << ghobj << std::endl;
-  if (dry_run)
-    return 0;
+  SnapSet ss;
+  if (ghobj.hobj.has_snapset()) {
+    r = get_snapset(store, coll, ghobj, ss, false);
+    if (r < 0) {
+      cerr << "Can't get snapset error " << cpp_strerror(r) << std::endl;
+      return r;
+    }
+    if (!ss.snaps.empty() && !all) {
+      if (force) {
+        cout << "WARNING: only removing "
+             << (ghobj.hobj.is_head() ? "head" : "snapdir")
+             << " with snapshots present" << std::endl;
+        ss.snaps.clear();
+      } else {
+        cerr << "Snapshots are present, use removeall to delete everything" << std::endl;
+        return -EINVAL;
+      }
+    }
+  }
+
   ObjectStore::Transaction t;
   OSDriver::OSTransaction _t(driver.get_transaction(&t));
-  r = mapper.remove_oid(ghobj.hobj, &_t);
-  if (r < 0 && r != -ENOENT) {
-    cerr << "remove_oid returned " << cpp_strerror(r) << std::endl;
-    return r;
+
+  cout << "remove " << ghobj << std::endl;
+
+  if (!dry_run) {
+    r = remove_object(coll, ghobj, mapper, &_t, &t);
+    if (r < 0)
+      return r;
+  }
+
+  ghobject_t snapobj = ghobj;
+  for (vector<snapid_t>::iterator i = ss.snaps.begin() ;
+       i != ss.snaps.end() ; ++i) {
+    snapobj.hobj.snap = *i;
+    cout << "remove " << snapobj << std::endl;
+    if (!dry_run) {
+      r = remove_object(coll, snapobj, mapper, &_t, &t);
+      if (r < 0)
+        return r;
+    }
   }
 
-  t.remove(coll, ghobj);
+  if (!dry_run)
+    store->apply_transaction(&osr, std::move(t));
```

Comment:
i think just checking for `dry_run` at this place would be enough. in other words, no need to check for `dry_run` every where.